index.js

```diff
@@ -1,171 +1,117 @@
-    // SDF. This is a significant performance win.
-    if (getMetrics && this.useMetrics) {
-        // The integer/pixel part of the top alignment is encoded in metrics.top
-        // The remainder is implicitly encoded in the rasterization
-        top = Math.floor(textMetrics.actualBoundingBoxAscent);
-        baselinePosition = this.buffer + Math.ceil(textMetrics.actualBoundingBoxAscent);
-        imgTop = this.buffer;
-        imgLeft = this.buffer;
-
-        // If the glyph overflows the canvas size, it will be clipped at the
-        // bottom/right
-        glyphWidth = Math.min(this.size,
-            Math.ceil(textMetrics.actualBoundingBoxRight - textMetrics.actualBoundingBoxLeft));
-        glyphHeight = Math.min(this.size - imgTop,
-            Math.ceil(textMetrics.actualBoundingBoxAscent + textMetrics.actualBoundingBoxDescent));
-
-        width = glyphWidth + doubleBuffer;
-        height = glyphHeight + doubleBuffer;
-        this.ctx.textBaseline = 'alphabetic';
-    } else {
-        width = glyphWidth = this.size;
-        height = glyphHeight = this.size;
-        // 19 points is an approximation of the "cap height" ascent from alphabetic
-        // baseline (even though actual drawing is from middle baseline, we can
-        // use the approximation because every glyph fills the em box)
-        top = 19 * this.fontSize / 24;
-        imgTop = imgLeft = 0;
-        baselinePosition = this.middle;
-        this.ctx.textBaseline = 'middle';
-    }
+        const data = new Uint8ClampedArray(width * height);
+        const glyph = {data, width, height, glyphWidth, glyphHeight, glyphTop, glyphLeft, glyphAdvance};
+        if (glyphWidth === 0 || glyphHeight === 0) return glyph;
 
-    var imgData;
-    if (glyphWidth && glyphHeight) {
-        this.ctx.clearRect(imgLeft, imgTop, glyphWidth, glyphHeight);
-        this.ctx.fillText(char, this.buffer, baselinePosition);
-        imgData = this.ctx.getImageData(imgLeft, imgTop, glyphWidth, glyphHeight);
-    }
+        const {ctx, buffer, gridInner, gridOuter} = this;
+        ctx.clearRect(buffer, buffer, glyphWidth, glyphHeight);
+        ctx.fillText(char, buffer, buffer + glyphTop + 1);
+        const imgData = ctx.getImageData(buffer, buffer, glyphWidth, glyphHeight);
 
-    var alphaChannel = new Uint8ClampedArray(width * height);
+        // Initialize grids outside the glyph range to alpha 0
+        gridOuter.fill(INF, 0, width * height);
+        gridInner.fill(0, 0, width * height);
 
-    prepareGrids(imgData, width, height, glyphWidth, glyphHeight, this.gridOuter, this.gridInner);
+        const offset = (width - glyphWidth) >> 1;
 
-    edt(this.gridOuter, width, height, this.f, this.v, this.z);
-    edt(this.gridInner, width, height, this.f, this.v, this.z);
+        for (let y = 0; y < glyphHeight; y++) {
+            for (let x = 0; x < glyphWidth; x++) {
+                const j = (y + offset) * width + x + offset;
```

Comment:
Did you go back to the original grid preparation code here mainly because it's more compact? I made the changes after experimenting, just reasoning that since it was such a tight loop small things could make a difference. But the idea was (1) half the number of branch instructions, and (2) simple multiplication may be cheaper than `Math.pow`. My benchmarking strategy was to profile a glyph-heavy load where TinySDF took about 500ms and look specifically at the time spent in prepareGrids. It ended up looking like a 15ms->10ms change over several runs. So pretty small and maybe not worthwhile overall. I can try running again to get the numbers documented.

Reply:
@ChrisLoer yeah, in my tests it didn't make much difference and made the code a lot more verbose, but I can backtrack on this now that the overall code is smaller.